### PR TITLE
Move YAML examples from integration tests to the demos, ensure all demos have test coverage

### PR DESCRIPTION
**Pull request by v1v**
## Description

Programatically validate the demos within the integrations module. This is a subset of what we aimed with https://github.com/jenkinsci/configuration-as-code-plugin/pull/621.

This is a very basic approach, so far I've decided to use the README.md with the YAML format itself to simplify the parsing. Ideally, the format under the README.md should be validated. The idea is to expose this idea and see whether it makes any sense any further developments.

Note, ideally `JenkinsConfiguredWithReadmeRule` should be refactored as it does almost the same as `JenkinsConfiguredWithCodeRule`.

### What's the idea?
- Define a standard demos/<plugin>/README.md convention
- Create integration tests with the `@JenkinsConfiguredWithReadmeRule` annotation.
- Avoid any kind of untested demos yaml

### Potentially improvements

We could potentially simplify the README.md format with the include directive but it seems it's not supported yet within the GitHub itself. See
- https://github.com/github/markup/issues/172
- https://github.com/github/markup/issues/346
- https://github.com/github/markup/issues/1095

### Highlights
- enable `@JenkinsConfiguredWithReadmeRule` annotation.
- Move some integrations/resources/**/*.yml files to the demos/**/README.md format with the fenced code block format.
- Parse markdown with the flexmark library.
- Bump mailer plugin dependency as requested.

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### Next actions
- This is just a framework with a few examples, let's move the remaining examples from the integrations into the demos in some follow-ups.
- Support multiple fenced code block entries within the README.md and the `@JenkinsConfiguredWithReadmeRule`.

### CODE REVIEW

**Comment by v1v** [comment]
Sorry for the spam, I really wanted to create a draft PR, but I somehow managed to create the ready one. 😿

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    //TODO: Looks like API defect, exception should be thrown
+    /**
+     * Exports the Jenkins configuration to a string.
+     * @return YAML as string
+     * @param strict Fail if any export operation returns error
+     * @throws Exception Export error
+     * @throws AssertionError Failed to export the configuration
+     * @since 1.25
+     */
+    public String exportToString(boolean strict) throws Exception {
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java]: should be strict by default in new code I assume?

File: demos/mailer/README.md [hunk 2]
```diff
@@ -1,24 +1,9 @@
...
-## sample configuration
-
-```yaml
+## Requires `mailer-plugin` >= 1.22
 unclassified:
   mailer:
     adminAddress: admin@acme.org
     replyToAddress: do-not-reply@acme.org
-    smtpHost: smtp.acme.org
+    #smtpHost: smtp.acme.org
```
[the remark this answers is not in the document]
Comment (casz) [reply demos/mailer/README.md]: You should also bump the minimum required for the demo!

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java (hunk 1 above)
Comment (v1v) [reply plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java]: I just refactored it: 
- https://github.com/jenkinsci/configuration-as-code-plugin/pull/1055/commits/cdde7b89c04b7aec10e62fd14f1f65f32d5f74ee

File: demos/mailer/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (v1v) [reply demos/mailer/README.md]: done, although it seems:

```yaml
unclassified:
  mailer:
    adminAddress:

```

causes some errors:

```
Caused by: io.jenkins.plugins.casc.ConfiguratorException: 'adminAddress' is deprecated
	at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:320)
	at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:287)
	at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:352)
	at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:287)
	at io.jenkins.plugins.casc.ConfigurationAsCode.lambda$checkWith$7(ConfigurationAsCode.java:709)
	at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:659)
	... 10 more
```

That's the reason I commented that particular configuration in the README.md

File: demos/mailer/README.md [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
-## Requires `mailer-plugin` >= 1.22
+Requires `mailer-plugin` >= 1.24
+
+## sample configuration
+
+```yaml
 unclassified:
   mailer:
-    adminAddress: admin@acme.org
+    # adminAddress: admin@acme.org          # This is deprecated
```
Comment (casz) [inline demos/mailer/README.md]: please remove as it have been moved to `unclassified: location: adminAddress`

File: demos/mailer/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (casz) [reply demos/mailer/README.md]: See https://github.com/jenkinsci/configuration-as-code-plugin/pull/1055#discussion_r324849029

File: demos/mailer/README.md [hunk 4]
```diff
@@ -1,24 +1,13 @@
 # mailer plugin
 
-Requires `mailer-plugin` >= 1.22
-
-Mailer plugin configuration belongs under `unclassified` root element
+Requires `mailer-plugin` >= 1.24
```
Comment (timja) [inline demos/mailer/README.md]:
```suggestion
Requires `mailer-plugin` >= 1.25
```

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ * @author v1v (Victor Martinez)
+ */
+public class JenkinsConfiguredWithReadmeRule extends JenkinsConfiguredRule {
+
+    static final DataHolder OPTIONS = PegdownOptionsAdapter.flexmarkOptions(
+        Extensions.ALL
+    );
+    static final Parser PARSER = Parser.builder(OPTIONS).build();
+
+
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java]:
```suggestion
```

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                    }
+                } else {
+                    throw new IllegalStateException(
+                        "Field must be public static JenkinsConfiguredWithReadmeRule");
+                }
+            }
+        }
+        return null;
+    }
+
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/JenkinsConfiguredWithReadmeRule.java]:
```suggestion
```

File: integrations/src/test/java/io/jenkins/plugins/casc/ActiveDirectoryTest.java [hunk 7]
```diff
@@ -21,10 +21,10 @@
     @Rule
     public RuleChain chain = RuleChain.outerRule(new EnvironmentVariables()
             .set("BIND_PASSWORD", "ADMIN123"))
-            .around(new JenkinsConfiguredWithCodeRule());
+            .around(new JenkinsConfiguredWithReadmeRule());
 
     @Test
-    @ConfiguredWithCode(value = "ActiveDirectoryTest.yml")
+    @ConfiguredWithReadme(value = "active-directory/README.md")
```
Comment (timja) [inline integrations/src/test/java/io/jenkins/plugins/casc/ActiveDirectoryTest.java]:
```suggestion
    @ConfiguredWithReadme("active-directory/README.md")
```

File: plugin/pom.xml [hunk 8]
```diff
@@ -70,6 +70,11 @@
       <version>1.19.0</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.vladsch.flexmark</groupId>
+      <artifactId>flexmark-all</artifactId>
+      <version>0.50.40</version>
```
Comment (casz) [inline plugin/pom.xml]: Not sure I like yet another dependency for our test package.
Perhaps we should consider splitting the test package and adding a way for us to declare dependencies for our test package

At least
```suggestion
      <version>0.50.40</version>
      <scope>test</scope>
```

File: demos/sonarqube/README.md [hunk 9]
```diff
@@ -2,26 +2,26 @@
 
 ## sample configuration
 
-Sample configuration for the SonarQube plugin.
+Sample configuration for the [SonarQube plugin](https://plugins.jenkins.io/sonar).
 
 *Below sample configuration willingly set all attributes values because of current issues with Sonar plugin 2.9 version. (cf. #982)*
 
-```yaml
+```yaml
```
Comment (casz) [inline demos/sonarqube/README.md]:
```suggestion
```yaml
```

File: plugin/pom.xml (hunk 8 above)
Comment (v1v) [reply plugin/pom.xml]: I initially used the test scope but I was struggling with the class loader issue when running with the new annotation, see the below stacktrace:

```
[ERROR] configure_sonar_globalconfig(io.jenkins.plugins.casc.SonarQubeTest)  Time elapsed: 0 s  <<< ERROR!
java.lang.NoClassDefFoundError: com/vladsch/flexmark/util/ast/Node
	at io.jenkins.plugins.casc.SonarQubeTest.<init>(SonarQubeTest.java:20)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:217)
	at org.junit.runners.BlockJUnit4ClassRunner$1.runReflectiveCall(BlockJUnit4ClassRunner.java:266)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:290)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.ClassNotFoundException: com.vladsch.flexmark.util.ast.Node
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 25 more
```

So I just managed to include the test dependency also in the integrations too:
- https://github.com/jenkinsci/configuration-as-code-plugin/pull/1055/commits/f7a0e986c6ac5ad542314d8de1f2d07617f9f4dd

**Comment by v1v** [comment]
@casz, not sure if it's required something else in my end. I'd like to move the remaining integration test cases with some follow-ups, as I didn't want a massive big PR with a bunch of changes.

Please let me know how to proceed if required.

Thanks

**Comment by casz** [comment]
Merged 🙌

Hopefully we can move forward with getting all the demos validated :)

**Comment by v1v** [comment]
❤️

**Comment by casz** [comment]
@v1v 
I am afraid with approach could make the integration/pom.xml may become un-maintainable

**Pull request merged**